django/db/models/fields/__init__.py

```diff
@@ -392,9 +392,11 @@ def _check_deprecation_details(self):
         return []
 
     def get_col(self, alias, output_field=None):
+        alternate_output_field = True
         if output_field is None:
             output_field = self
-        if alias != self.model._meta.db_table or output_field != self:
+            alternate_output_field = False
+        if alias != self.model._meta.db_table or (alternate_output_field and output_field != self):
```

Comment:
As I alluded on the ticket, readability/clarity was a concern when I wrote it, and vs. the other option suggested by Carlton, without making any judgement about better/worse. So I'm absolutely on board with keeping it easy.

I'm fine with making the change proposed above or in the ticket, on the basis that the pair of you are fare more likely to have to revisit it than I, and a lot of this kind of thing is subjective, but I do (with respect, which I appreciate doesn't come across in text well!) disagree that the above is more readable, on the following basis:
- `source_output_field` is an odd name, because yes, the source has an intrinsic output field, so you might expect (imagine the method grows) that `source_output_field` is `self` rather than a boolean. `use_source_as_output_field` would be more appropriate (wordiness notwithstanding)
- `source_output_field = output_field is None` and `if source_output_field:` and `not source_output_field` means 2 additional truthiness tests instead of 1. All of those are about as quick as you're going to get using Python, but still.
- `or (not x and y != z)` is not easy to parse IMHO because of the negatives and inversion of boolean logic. But that's probably just my intelligence failing to show up to the party :)

In case you haven't had the opportunity to cast your eye over it on the ticket, Carlton's alternate version was:
```
if alias != self.model._meta.db_table or (output_field and output_field != self):
        from django.db.models.expressions import Col
        return Col(alias, self, output_field or self)
    else:
        return self.cached_col
```
It was indicated that deferral to you in review, so I just wanted to make sure you have all the options and trade-offs.

Reply:
> * `source_output_field` is an odd name, because yes, the source has an intrinsic output field, so you might expect (imagine the method grows) that `source_output_field` is `self` rather than a boolean. `use_source_as_output_field` would be more appropriate (wordiness notwithstanding)

Agreed `use_source_as_output_field` or even `use_base_output_field` sounds better.

> * `or (not x and y != z)` is not easy to parse IMHO because of the negatives and inversion of boolean logic. But that's probably just my intelligence failing to show up to the party :)

Maybe it's harder to parse but (IMO) it easier to read and understand :wink: : _`If we don't use the base output field and the passed output field differs from the base one then ... `_ Currently the negation is hidden in the variable name `alternate_...` :shrug: